kore/src/Kore/Rewrite/Function/Evaluator.hs

```diff
@@ -191,6 +201,88 @@ evaluatePattern
             sideCondition
             & maybeT (defaultValue Nothing) return
 
+lookupAxiomSimplifier ::
+    MonadSimplify simplifier =>
+    TermLike RewritingVariableName ->
+    MaybeT simplifier BuiltinAndAxiomSimplifier
+lookupAxiomSimplifier termLike = do
+    simplifierMap <- lift askSimplifierAxioms
+    hookedSymbols <- lift askHookedSymbols
+    let missing = do
+            -- TODO (thomas.tuegel): Factor out a second function evaluator and
+            -- remove this check. At startup, the definition's rules are
+            -- simplified using Matching Logic only (no function
+            -- evaluation). During this stage, all the hooks are expected to be
+            -- missing, so that is not an error. If any function evaluators are
+            -- present, we assume that startup is finished, but we should really
+            -- have a separate evaluator for startup.
+            Monad.guard (not $ null simplifierMap)
+            case termLike of
+                App_ symbol _
+                    | isDeclaredFunction symbol -> do
+                        let hooked = criticalMissingHook symbol
+                            unhooked = warnFunctionWithoutEvaluators symbol
+                        maybe unhooked hooked $ getHook symbol
+                _ -> return ()
+            empty
+    maybe missing return $ do
+        axiomIdentifier <- Axiom.Identifier.matchAxiomIdentifier termLike
+        let exact = Map.lookup axiomIdentifier simplifierMap
+        case axiomIdentifier of
+            Axiom.Identifier.Application appId ->
+                let builtintEvaluator = do
+                        name <- Map.lookup appId hookedSymbols
+                        koreEvaluators name
+                 in combineEvaluatorsWithFallBack (builtintEvaluator, exact)
```

Comment:
There is another optimization we could potentially do. If we match on TermLike here, if it is an `ApplySymbolF`, it already carries a `Hook` and we can potentially bypass a lookup inside `hookedSymbols`? Not sure if this is sound or worth it, just thought I'd mention it:

```haskell
ApplySymbolF (Application Symbol{hook} _) | Just name <- getHook hook -> 
  combineEvaluatorsWithFallBack (koreEvaluators name, exact)
```

Reply:
Hm good point! I see that, in `matchAxiomIdentifier`, we use [`mkAppId`](https://github.com/runtimeverification/haskell-backend/blob/ac0e3c0c403bbca4d812d921f8783f4dfcdbcb29/kore/src/Kore/Rewrite/Axiom/Identifier.hs#L202) to extract an `AxiomIdentifier` from an `Application`. Instead, we could just extract the hook directly from the `Application`.

I see that `mkAppId` is also used in `InjF`, `EndiannessF` and `SignednessF`. Would it make sense to generalize this as below?

```hs
    getHookFromTermLikeF :: TermLikeF RewritingVariableName a -> Maybe Text
    getHookFromTermLikeF = \case
        ApplySymbolF (Application symbol _) -> getHook symbol
        InjF (Inj.toApplication -> Application symbol _) -> getHook symbol
        EndiannessF (getConst -> Endianness.toApplication -> Application symbol _) -> getHook symbol 
        SignednessF (getConst -> Signedness.toApplication -> Application symbol _) -> getHook symbol 
        _ -> Nothing
```

Reply:
@ana-pantilie?

Reply:
I'm not sure, you could give it a try, but I'd prefer to get this merged first and then open a follow-up PR which attempts these changes.

Reply:
Sure! @goodlyrottenapple / @jberthold can you please review/approve this PR?

Reply:
I've opened a separate PR with this change: #3195 

Please see this comment: https://github.com/runtimeverification/haskell-backend/pull/3195#issuecomment-1214425441